src/uct/ib/base/ib_md.c

```diff
@@ -923,13 +891,53 @@ uct_ib_mkey_pack(uct_md_h uct_md, uct_mem_h uct_memh,
             }
         }
 
-        mkey = memh->exported_lkey;
-    } else if ((flags & UCT_MD_MKEY_PACK_FLAG_INVALIDATE_RMA) &&
-               ((atomic_rkey != UCT_IB_INVALID_MKEY) ||
-                !(memh->flags & UCT_IB_MEM_ACCESS_REMOTE_ATOMIC))) {
-        /**
+        uct_ib_md_pack_exported_mkey(md, memh->exported_lkey, mkey_buffer);
+        return UCS_OK;
+    }
+
+    /* Create an atomic key only if a user requested atomic access to the
+     * memory region and the hardware supports it.
+     */
+    if (memh->flags & UCT_IB_MEM_ACCESS_REMOTE_ATOMIC) {
+        if (!(memh->flags & UCT_IB_MEM_FLAG_ATOMIC_MR)) {
+            /* create UMR on-demand */
+            status = UCS_PROFILE_NAMED_CALL_ALWAYS("reg atomic key",
+                                                   md->ops->reg_atomic_key, md,
+                                                   memh);
+            if (status == UCS_OK) {
+                memh->flags |= UCT_IB_MEM_FLAG_ATOMIC_MR;
+                extra_rkey   = memh->atomic_rkey;
+                ucs_trace("created atomic key 0x%x for 0x%x", memh->atomic_rkey,
+                          memh->lkey);
+            } else if (status == UCS_ERR_UNSUPPORTED) {
+                if (flags & UCT_MD_MKEY_PACK_FLAG_INVALIDATE_AMO) {
+                    /* Must have indirect atomic key to invalidate */
+                    return UCS_ERR_IO_ERROR;
+                }
+
```

Comment:
maybe explicitly set `extra_rkey = UCT_IB_INVALID_MKEY` here for documentation purposes? (instead of line 925)